### PR TITLE
fix Stage state transitionToRunning on schedulingComplete()

### PR DESCRIPTION
**Pull request by guhanjie**
fix Stage state transitionToRunning on schedulingComplete(), origin is dead code.

It would be possible to make Stage transition to RUNNING on schedulingComplete(), 
which maybe earlier than StageTaskListener#stateChanged()

### CODE REVIEW

File: core/trino-main/src/main/java/io/trino/execution/SqlStageExecution.java
Comment (guhanjie) [reply core/trino-main/src/main/java/io/trino/execution/SqlStageExecution.java:255]: > @sopel39 the code before the change clearly uses a wrong condition, since was calling `this.getState()` for each task.
> 
> @guhanjie the transition seems to happen here
> https://github.com/trinodb/trino/blob/77a24a07beed1752014d28ec7cc4294ab4875d36/core/trino-main/src/main/java/io/trino/execution/SqlStageExecution.java#L545
> 
> maybe instead of fixing this line, we can remove it? (thinking loud)
> 
> @dain ptal

yes, there are two points to RUNNING, one is we talking about, the other one is you mentioned that, which can make sure the Stage state transition to RUNNING. Actually, all achieve to RUNNING is go through this way

Comment (findepi) [reply core/trino-main/src/main/java/io/trino/execution/SqlStageExecution.java:255]: @guhanjie i understand you found the bug by looking at the code -- via code inspection. 
the code should be fixed, the question is -- what should be the fix.

if we're changing the code to make the condition be applicable, we should ask ourselves, what problem are we solving from end-user perspective. ie. is there a functional bug?
I don't feel safe proceeding without answer to that question, and i don't know the answer.

If we remove the code as dead code  -- as is currently is -- we don't make a behavioral difference while we also improve readability, and that's something i would approve right away.

**Review by sopel39 — APPROVED** [review]
I've approved (since it was obviously wrong), but it would be great if another person took a look (@dain ?)

Seems like this line never worked, so it might be not needed. Yet maybe it's intention was to reduce state change latency

File: core/trino-main/src/main/java/io/trino/execution/SqlStageExecution.java
Comment (guhanjie) [reply core/trino-main/src/main/java/io/trino/execution/SqlStageExecution.java]: agree with you, i removed the code(never worked currently), it is safe and can reduce the tasks iteration checks and code is clear(there is only one point to RUNNING)
and as i mentioned above, i also think the state change latency would be no impact, 
because the first point faster than the second point obviously is such rare in fact.

I'm pleasure to see more looks or talks before the change code is approved. @dain  @sopel39

**Review by sopel39 — APPROVED** [review]

**Review by findepi — APPROVED** [review]

**Comment by findepi** [comment]
@guhanjie please squash commits and add a note why we considered this "dead code", so that someone looking at the commit later will understand the change easily.

**Comment by guhanjie** [comment]
> @guhanjie please squash commits and add a note why we considered this "dead code", so that someone looking at the commit later will understand the change easily.

I have no permission to Merge pull request, i have to rebase locally and force push it 😭 
Could anyone authorize me the permission??

image

**Comment by sopel39** [comment]
> Could anyone authorize me the permission??

Only maintainers group members can merge pull requests.

**Comment by sopel39** [comment]
Thanks for contribution!

**Pull request merged**
